addons/social_share/models/campaign.py

```diff
@@ -0,0 +1,230 @@
+                    f'{element.role}_color': element.text_color
+                    for element in campaign.element_ids
+                },
+                raise_on_code=False,
+            ) if campaign.template_id.body else '' for campaign in self
+        ]
+
+    def _get_images_bytes(self, record=None, with_demo_values=False):
+        bodies = self._get_bodies(record=record, with_demo_values=with_demo_values)
+        if self and bodies:
+            return self.env['ir.actions.report']._run_wkhtmltoimage(
+                bodies,
+                *TEMPLATE_DIMENSIONS
+            )
+        return [b''] * len(self)
+
+    def _get_images_b64(self, record=None, with_demo_values=False):
+        return [
+            base64.b64encode(image_bytes) for image_bytes
+            in self._get_images_bytes(record=record, with_demo_values=with_demo_values)
+        ]
+
+    @api.onchange('template_group_id')
+    def _onchange_template_group_id(self):
+        for campaign in self:
+            campaign.template_id = campaign.template_group_id.template_ids[:1]
+
+    def action_open_url_share(self):
+        """Open url dialog."""
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Share Link'),
+            'res_model': 'snshare.url.share',
+            'views': [[False, 'form']],
+            'context': {
+                'default_campaign_id': self.id,
+                'dialog_size': 'medium',
+            },
+            'target': 'new',
+        }
+
+    def action_show_clicked_urls(self):
+        self.ensure_one()
+        return self.env["ir.actions.actions"]._for_xml_id("social_share.action_snshare_url") | {
+            'context': {'search_default_filter_visited': True},
+            'domain': [('campaign_id', '=', self.id)],
+        }
+
+    def action_show_shared_urls(self):
+        self.ensure_one()
+        return self.env["ir.actions.actions"]._for_xml_id("social_share.action_snshare_url") | {
+            'context': {'search_default_filter_shared': True},
+            'domain': [('campaign_id', '=', self.id)],
+        }
+
+    def action_share_multi(self):
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Send Share URLS'),
+            'res_model': 'snshare.url.share.multi',
+            'context': {'default_share_campaign_id': self.id, 'default_subject': self.name},
+            'views': [[False, 'form']],
+            'target': 'new',
+        }
+
+    def _get_url(self, record_id):
+        self.ensure_one()
+        uuid = False
+        if self.model_id:
+            existing_url = self.share_url_ids.filtered(lambda rec: rec.res_id == record_id)
+            if existing_url:
+                uuid = existing_url.uuid
+            else:
+                uuid = self.env['snshare.url'].create({'campaign_id': self.id, 'res_id': record_id}).uuid
+        return SocialShareController._get_campaign_url(
```

Comment:
Should we instead put this method in this file? And call it from the controller / the wizard? Does it not feel strange to call it on the controller while being on the campaign model? :p